### ISSUE TITLE
SignalAndWait is not implemented for Unix

### ISSUE THREAD

**Issue by danmosemsft** [dotnet/coreclr#10441]
We believe this is possible to implement for Unix but it's a low priority as the API is flawed and it has extremely low usage.

https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/WaitHandle.cs#L451
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/WaitHandle.cs#L464
https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Threading/WaitHandle.cs#L483

====
> @kouvel SignalAndWait is a bit complicated because it has to work on any WaitHandle created by the user, which can have any handle value. So it’s not as simple as just calling the respective Signal method followed by the Wait method, instead we would have to implement this in PAL and determined that it’s not worth doing at the time.

>However, we could potentially do this as a simple compromise:
-	Check the type of the wait handle to signal
o	If it’s an EventWaitHandle, call SetEvent
o	If it’s a Semaphore, call Release
o	If it’s a Mutex, call ReleaseMutex
o	If it’s none of the above, throw PNSE
-	Wait on the handle to wait on

>Users can still change the handle of a Mutex/Semaphore/EventWaitHandle object, which is bizarre, but they would get the appropriate behavior of the signal methods in that case.
----
@jkotas 
>It is not that much harder to do switch on the handle type in the PAL to avoid the problem you have mentioned… .
>I do not have a much opinion on this one. I will leave it up to you to decide what to do about this one. It is somewhat broken API even on Windows because of it does not work on STA threads (for no good reason). I would expect it to be used very rarely because of that.
----
@kouvel 
>Agreed, it’s not far off from just implementing it properly in the PAL.

>I’ll discuss with Gaurav. I plan on implementing it properly for CoreRT Unix since there are no such issues there, so maybe for consistency it would make sense to implement it in CoreCLR as well.

**Comment by kouvel** [comment]
Could be pushed to Future if necessary, but putting it in 2.0.0 for now since it would be nice to get it done in that timeframe

**Comment by evil-shrike** [comment]
Hi, is there any news on `SignalAndWait` implementation on Linux?
Is it a correct workaround to replace `SignalAndWait (toSignal, toWait, Timeout, false)` with 
```
    toSignal.Set();
    bool result = toWait.WaitOne(Timeout, false);
```
?

**Comment by kouvel** [comment]
I'm planning on adding this for 2.1. Yes the above is a close equivalent to what SignalAndWait would do.
